### PR TITLE
Update to Kotlin 1.4 and bump version number to 1.0.0

### PR DESCRIPTION
**Pull request by antoinelamy**
## Description
- Update Kotlin compiler to `1.4.0`
- Remove `GRADLE_METADATA` flag as module metadata is used in dependency resolution and included in publications by default in Gradle 6.0 and above
- Remove manual `stdlib` dependencies as they are now included by default
- Migrate to hierarchical project structure and use platform shortcuts.
- Bump version number to `1.0.0` and migrate to semantic versioning
- Move away from the object based DispatchQueue on iOS and favour GCD instead of NSOperationQueue
- Create a serial queue on iOS to use with subscriptions
- Update deprecated reference to`kotlin.browser.window`

## Motivation and Context
As Kotlin 1.4 has been officially released in the stable channel, it is time to update.

## How Has This Been Tested?
Tests in place still pass, to be tested integrated into an existing project.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

Every project using Trikot.Foundation will need to update their own version of Kotlin to get the latest updates.

### CODE REVIEW

File: trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSGlobalDispatchQueue.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.mirego.trikot.foundation.concurrent.dispatchQueue
+
+import com.mirego.trikot.foundation.concurrent.freeze
+import platform.darwin.DISPATCH_QUEUE_PRIORITY_HIGH
+import platform.darwin.dispatch_async
+import platform.darwin.dispatch_get_global_queue
+
+open class iOSGlobalDispatchQueue : DispatchQueue {
```
Comment (npresseault) [inline trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSGlobalDispatchQueue.kt:8]: `IOSGlobalDispatchQueue`

File: trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSSerialDispatchQueue.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package com.mirego.trikot.foundation.concurrent.dispatchQueue
+
+import com.mirego.trikot.foundation.concurrent.freeze
+import platform.darwin.dispatch_async
+import platform.darwin.dispatch_queue_create
+import platform.darwin.dispatch_queue_serial_t
+
+open class iOSSerialDispatchQueue(identifier: String) : DispatchQueue {
```
Comment (npresseault) [inline trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSSerialDispatchQueue.kt:8]: `IOSSerialDispatchQueue`

File: trikotFoundation/src/iosMain/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSGlobalDispatchQueue.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package com.mirego.trikot.foundation.concurrent.dispatchQueue
+
+import com.mirego.trikot.foundation.concurrent.freeze
+import platform.darwin.DISPATCH_QUEUE_PRIORITY_HIGH
+import platform.darwin.dispatch_async
+import platform.darwin.dispatch_get_global_queue
+
+open class iOSGlobalDispatchQueue : DispatchQueue {
```
Comment (npresseault) [inline trikotFoundation/src/iosMain/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSGlobalDispatchQueue.kt:8]: `IOSGlobalDispatchQueue`

**Comment by flambert** [comment]
Why the apparent code duplication for tvosMain?

File: trikotFoundation/src/iosMain/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSSerialDispatchQueue.kt [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package com.mirego.trikot.foundation.concurrent.dispatchQueue
+
+import com.mirego.trikot.foundation.concurrent.freeze
+import platform.darwin.dispatch_async
+import platform.darwin.dispatch_queue_create
+import platform.darwin.dispatch_queue_serial_t
+
+open class iOSSerialDispatchQueue(identifier: String) : DispatchQueue {
```
Comment (npresseault) [inline trikotFoundation/src/iosMain/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSSerialDispatchQueue.kt:8]: `IOSSerialDispatchQueue`

File: trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSGlobalDispatchQueue.kt (hunk 1 above)
Comment (npresseault) [reply trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSGlobalDispatchQueue.kt:8]: What's different about arm32 that we can't use the same code? The long types?

File: trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/DispatchQueues.kt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package com.mirego.trikot.foundation.concurrent.dispatchQueue
+
+actual class OperationDispatchQueue : iOSGlobalDispatchQueue(), DispatchQueue
+
+actual class SerialSubscriptionDispatchQueue : iOSSerialDispatchQueue("serialSubscriptionQueue"), DispatchQueue
```
Comment (martingagnon) [inline trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/DispatchQueues.kt:5]: I think we could deprecate this one. The serialDispatchQueue was introduce as a way to deal with concurrency. A SerialQueue was added in the publishSubject to manage this concurrency.

File: trikotFoundation/build.gradle [hunk 6]
```diff
@@ -27,20 +27,11 @@ repositories {
 group 'com.mirego.trikot'
 
 kotlin {
-    android {
-        publishAllLibraryVariants()
```
Comment (martingagnon) [inline trikotFoundation/build.gradle:31]: Not required anymore ? :)

File: trikotFoundation/build.gradle [hunk 7]
```diff
@@ -27,20 +27,11 @@ repositories {
...
-        fromPreset(presets.jvm, 'jvm')
-        fromPreset(presets.android, 'android')
-        fromPreset(presets.js, 'js')
-        fromPreset(presets.iosArm32, 'iosArm32')
-        fromPreset(presets.iosArm64, 'iosArm64')
-        fromPreset(presets.iosX64, 'iosX64')
-        fromPreset(presets.tvosArm64, 'tvosArm64')
-        fromPreset(presets.tvosX64, 'tvosX64')
-    }
+    android()
```
Comment (martingagnon) [inline trikotFoundation/build.gradle:30]: No JVM targets ?

**Comment by antoinelamy** [comment]
> Why the apparent code duplication for tvosMain?

From the Kotlin documentation:
> You can use platform-dependent libraries like Foundation, UIKit, and POSIX in source sets shared among several native targets. This helps you share more native code without being limited by platform-specific dependencies.
>
> No additional steps are required – everything is done automatically. IntelliJ IDEA will help you detect common declarations that you can use in the shared code.
>
> However, note that there are some limitations:
>
> This approach works only for a native source set that is shared among platform-specific source sets. It doesn’t work for native source sets shared at higher levels of the source set hierarchy.
> - For example, if you have `nativeDarwinMain` that is a parent of `watchosMain` and `iosMain`, where `iosMain` has two children – `iosArm64Main` and `iosX64Main`, you can use platform-dependent libraries only for `iosMain` but not for `nativeDarwinMain`.
> - It works only for interop libraries shipped with Kotlin/Native.

In short, a source set that is shared between iOS and tvOS cannot import platform specific libraries. Sad to duplicate but doesn't any other way around.

File: trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSGlobalDispatchQueue.kt (hunk 1 above)
Comment (antoinelamy) [reply trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSGlobalDispatchQueue.kt:8]: Exactly, the long flag variable 🙄

**Comment by marclefrancois** [comment]
In reference with @npresseault comments, I know we normally never capitalize de `i` in `iOS` but kinda feels weird to have a Kotlin class start with a lowercase. We really are in the details here but I think we should settle this clearly so we keep a consistent pattern everywhere. 

I do not have a strong opinion either way but the kotlin convention is (like all other language that I can think of) to start class names with a capital letter https://calendar.google.com/calendar/u/1/r

File: trikotFoundation/src/iosMain/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSGlobalDispatchQueue.kt (hunk 3 above)
Comment (antoinelamy) [reply trikotFoundation/src/iosMain/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSGlobalDispatchQueue.kt]: Done

File: trikotFoundation/src/iosMain/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSSerialDispatchQueue.kt (hunk 4 above)
Comment (antoinelamy) [reply trikotFoundation/src/iosMain/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSSerialDispatchQueue.kt]: Done

File: trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSSerialDispatchQueue.kt (hunk 2 above)
Comment (antoinelamy) [reply trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/iOSSerialDispatchQueue.kt]: Done

File: trikotFoundation/build.gradle (hunk 7 above)
Comment (antoinelamy) [reply trikotFoundation/build.gradle]: Oups, my mistake. Fixed!

File: trikotFoundation/build.gradle (hunk 6 above)
Comment (antoinelamy) [reply trikotFoundation/build.gradle:31]: Apparently yes, I put it back.

File: trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/DispatchQueues.kt (hunk 5 above)
Comment (antoinelamy) [reply trikotFoundation/src/iosArm32Main/kotlin/com/mirego/trikot/foundation/concurrent/dispatchQueue/DispatchQueues.kt]: Added a deprecated annotation

**Comment by antoinelamy** [comment]
After digging in Kotlin documentation, I found out that you can create a symlink to avoid code duplication so I took this approach for tvosMain

File: trikotFoundation/build.gradle [hunk 8]
```diff
@@ -1,21 +1,11 @@
...
     id 'com.android.library'
+    id 'kotlin-multiplatform'
     id 'org.jlleitschuh.gradle.ktlint'
     id 'mirego.release' version '1.13'
     id 'mirego.publish' version '1.0'
 }
 
-android {
-    compileSdkVersion 29
-    buildToolsVersion "29.0.3"
```
Comment (antoinelamy) [inline trikotFoundation/build.gradle:11]: Removed build tools version as it is no longuer needed as per documentation:
> You no longer need to specify a version for the build tools (so, you can now remove the android.buildToolsVersion property). By default, the plugin automatically uses the minimum required build tools version for the version of Android plugin you're using.

**Comment by flambert** [comment]
I usually use IOS and TVOS when the OS name is the suffix or the prefix. A random lowercase looks bad and is really not standard when it's a prefix!

**Comment by antoinelamy** [comment]
> I usually use IOS and TVOS when the OS name is the suffix or the prefix. A random lowercase looks bad and is really not standard when it's a prefix!

@flambert The prefix has already been renamed to `IOS` and because I removed tvos code duplication by creating a symlink, the `TVOS` prefix is not needed anymore.

**Comment by flambert** [comment]
> The prefix has already been renamed to `IOS` and because I removed tvos code duplication by creating a symlink, the `TVOS` prefix is not needed anymore.

I know, I'm only agreeing and adding more info from my perspective.

**Pull request merged**